### ISSUE TITLE
Get rid of ta-lib

### ISSUE THREAD

**Issue by hroff-1902** [freqtrade/freqtrade#2509]
Let's remove ta-lib from dependencies. This will simplify the installation process a lot.

* It's only used in default/sample strategy/hyperopt...
* A consensus for replacement library should be found.
* The doc part (installation instructions, examples of usage) can be moved to a separate doc file, tips for other TA libraries can also be collected there.
* How to deal with freqtrade/freqtrade-strategies?

A a way to do it, we could add `sample-<libname>-strategy.py` and ask users to test it for some period of time.

**Comment by xmatthias** [comment]
I fully agree with this - especially if you consider that the underlying C library has not been updated since 2013 according to their SVN trunk (talk about outdated libraries ...).

Removing it from the sample / default strategies should not be difficult (i'll can do that) and will fully support #2112 (it's even a point in that issue...).

Now i think that we'll need to keep it as separate dependency (`pip install freqtrade[talib]` ?) and install it in the docker-images for a while to facilitate a smooth transition.

we should be able to update most strategies in `freqtrade/freqtrade-strategies` to use the newly selected library (assuming we pick a library which supports most indicators) - and we can then deal the "remainders" separetely.

The difficult part is to find / agree on a future library.
You seem inclined towards finta - but really any library that supports "many" indicators will work, and we can add missing ones to technical (i would not start rebuilding such a library within technical though).

Maybe we can have `technical` depend on that selected library, import indicators from that into a specific area acting as "proxy" ... (i would not create a function for each, but directly import them as `from finta import *` .
Users could then use `import technical.indicators as ti` and have all indicators available as `ti.****()`.

The purpose of this can be so users only need to transition to technical - and should we find a "better" library, we can switch to that "transparently" for the user (not sure if this is a good thing though - since it can obscure if the indicator comes from finta, or from technical, or from another library).

**Comment by xmatthias** [comment]
Having tried to replace talib with finta - finta does certain calculations quite different.
Just try to `import finta as TA`, replace occurences of `ta.` with `TA.` (best one-by-one - to identify which indicators act differently) and try to run tests ...

certain tests fail after replacing certain indicators (rsi iirc)... but I've not yet had the time (or patience) to dig into the differences deeper.

**Comment by hroff-1902** [comment]
`from finta import TA` is the canonical way to import (as described in the finta's docs)
the indicators are class methods of finta.TA, not in the finta module itself.
(I did not say it fixes the problems, just a first reply into the comment above, started looking at this in details)

**Comment by hroff-1902** [comment]
finta is also broken in some places...

seems there is no /mature/ ta library in python out of the box...

**Comment by hroff-1902** [comment]
We have very limited choice:
https://libraries.io/github/search?keywords=technical-analysis&language=Python&sort=rank

**Comment by michielz** [comment]
Bitfinex made one for us, the code is ugly but it works well and is easy to use.
https://github.com/bitfinexcom/bfx-hf-indicators-py

I can confirm that the output for RSI, MACD, and EMA are correct when compared to TradingView's indicators.

To calculate the ATR like TradingView replace the return statement in `atr.py`:


    def tr(prevCandle, candle):
        if prevCandle == None:
          return 0

        # original code block
        # return max([
        #   prevCandle['high'] - prevCandle['low'],
        #   abs(candle['high'] - prevCandle['close']),
        #   abs(candle['low'] - prevCandle['close'])
        # ])

       # alternative code block, this is the way Tradingview calculates ATR
       return max([
         candle['high'] - candle['low'],  # changed line
         abs(candle['high'] - prevCandle['close']),
         abs(candle['low'] - prevCandle['close'])
       ])

**Comment by hroff-1902** [comment]
@michielz interesting, thanks!

**Comment by hroff-1902** [comment]
Another opportunity: tulip indicators library.
https://github.com/TulipCharts/tulipindicators
https://tulipindicators.org

Python bindings: https://github.com/cirla/tulipy

**Comment by zjhmale** [comment]
the code style of bfx-hf-indicators is not that ugly but it just uses the incremental way to do all the calculations for the sake of eliminating list operation overhead.
along the way, I think some indicators output by finta is not correct, e.g. EMA, etc.

**Comment by hroff-1902** [comment]
https://btalib.backtrader.com: a ta library which 
* pays attention to ta-lib (a ta-lib compatibility mode which can be switched on, see https://btalib.backtrader.com/ta-lib-compat/)
* seems to be well supported (last release is 2 days ago)
* design goals explicitly expressed: https://btalib.backtrader.com/introduction/#design

**Comment by xmatthias** [comment]
Just to throw in another alternative - https://github.com/bukosabino/ta

How about:
we decide on one of the alternatives (finta / ta / bta-lib / ...) for the default samples (samples, hyperopt, ...).

However, we also provide a overview page of additional TA libraries so users can decide depending on their needs (every library will have benefits / drawbacks, and we'll probably not find one library which as all indicators (also ta-lib does not have all indicators...)).

Additionally, we'd provide a "rather simple" method to install these dependencies into the docker image (or add them out of the box, most of the libraries mentioned here don't seem to occupy too much space).

**Comment by magick93** [comment]
HI @xmatthias - I like this approach. Basically have a 'reference implementation'. Or, to put it another way, rather than selecting one over another, architect a solution that allows to use each with minimal changes.

**Comment by MLidstrom** [comment]
Hi @xmatthias. Just started looking into Freqtrade and I got it up and running. Sort of. I ran into TA-Lib of course and I'm stuck. So I started looking for alternatives and found [Pandas TA](https://twopirllc.github.io/pandas-ta/). Looks impressive at first glance. What do you think? A possible candidate?

**Comment by xmatthias** [comment]
there's too many options, really :D all have certain "special" indicators - but all have something slightly diferent, too.

I'm not too big a fan of tools that modify / extend  the DataFrame interface ... as that makes the code sometimes quite difficult to digest (it's not clear why this works, but doesn't work the next try because the import is missing) - but that's not excluding anything.

To be honest though, i'm currently not actively looking at this, as it's working (although installation is a pain, sometimes) - and it's one of the best libraries when you look at performance alone.

**Comment by MLidstrom** [comment]
Congratulations to those that get it to work then :) I've been struggling with this for two days now and running out of options.

**Comment by xmatthias** [comment]
well it'll depend on the python you're using - this wheel will only work with Python 3.8. If you're using python3.7, you need to use the "other" Wheel - and if it's 3.6, you'll need to get the wheel yourself from the webpage in the ta-lib documentation).

**Comment by MLidstrom** [comment]
I have Python 3.8.5 installed. I've seen on various forums that Windows and TA-Lib isn't a good match so I'll probably experiment with Pandas TA.

**Comment by xmatthias** [comment]
well you can install it as "raw" python project too (`pip install ta-lib`) - but it'll require compiler dependencies.

If the above wheel does not match your file - then you'll most likely need the 32bit version (you can get this from [here](https://www.lfd.uci.edu/~gohlke/pythonlibs/#ta-lib) - as indicated in the ta-lib python wrapper documentation).

**Comment by MLidstrom** [comment]
I'm a senior programmer so I got the dependencies needed. I think. What I don't have is the patience 😄 I'll try the 32 bit version. Thank you!

**Comment by twopirllc** [comment]
Hello all!

Just came across **freqtrade**, looks 😎 , as well as found this issue to consider replacing the _de facto_ **TA Lib** Issue. I am the author of [**Pandas TA**](https://github.com/twopirllc/pandas-ta) library. Perhaps consider _Pandas TA_ as an optional or backup library if the user does not have _TA Lib_ installed or whichever package you end up choosing as _TA Lib_'s replacement.

One of the major reasons I wrote _Pandas TA_ was because _TA Lib_ is sometimes difficult to install (as previously noted). So _Pandas TA_ is a good replacement for _TA Lib,_ but does lack many of the Candle Stick Patterns so far. Also any shared indicators between _Pandas TA_ and _TA Lib_ (or TradingView) are 99% or more correlated to ensure consistent accuracy. Plus it has many features not available in similar libraries to help users easily build Strategies and run them easily.

So instead of rattling on about _Pandas TA_, I just want to say that I am around _if_ you have any further questions about it, how it works, or whatever... just @ me. Here to help the Algorithmic Trading Community.

Keep up the good work on **freqtrade**! 🥂 

Thanks for the consideration,
KJ

**Comment by xmatthias** [comment]
Thanks a lot for your suggestion - we'll make sure to consider it - one main point is for sure the activity (project is actively maintained) - which pandas ta definitely looks like :+1: 

There's no urgent need for us to replace ta-lib (as you can see by the age of this issue) - but if we're honest, only the python bindings of ta-lib are maintained, the underlying C code is barely findable and i don't think i've seen an update to it in the last 5 or so years (but maybe i'm looking in the wrong place ...)

**Comment by twopirllc** [comment]
@xmatthias,

> Thanks a lot for your suggestion and offer - we'll definitely consider this - one main point is for sure the activity (project is actively maintained) - which pandas ta definitely looks like 👍

Yeah, it will still be maintained for quite some time as I still have other things beyond just technical indicators I want to incorporate into the package. For instance, I have been silently incorporating some backtesting functionality into the package including _drawdown_ and common metrics like:

<img width="208" alt="Screen Shot 2020-11-15 at 9 20 59 AM" src="https://user-images.githubusercontent.com/18198392/99192124-fef9f780-2725-11eb-8dfa-1a23f95bbbde.png">

> There's no urgent need for us to replace ta-lib (as you can see by the age of this issue)

Of course.

> but if we're honest, only the python bindings of ta-lib are maintained, the underlying C code is barely findable and i don't think i've seen an update to it in the last 5 or so years (but maybe i'm looking in the wrong place ...)

Yes the Python wrappers are being maintained. You are correct that the underlying C code has not been modified publicly since 2013 from what I have found on their [SourceForge TA Lib](https://sourceforge.net/p/ta-lib/code/HEAD/tree/). Furthermore, _pyti_ has also stalled about 3 years ago as well. There are only a few Python TA Libraries that are actively maintained.

KJ

**Comment by LuisAlejandro** [comment]
Any of the proposed replacements has candlestick pattern recognition as TA-lib? I've been reviewing the ones that have been posted here without any luck.

**Comment by xmatthias** [comment]
If we (ever) replace ta-lib - then we'll only replace it as required dependency. 

Nothing prevents you from installing that yourself and continue using it in your strategy - so i don't really see a problem with that.

**Comment by LuisAlejandro** [comment]
Yes, I know, I'm also preparing to get rid of ta-lib on my strategies, but I don't seem to find a replacement for candlestick patterns, I figured this was the place for asking for help. Sorry if I'm disturbing the purpose of the ticket.

**Comment by xmatthias** [comment]
i think you'll find it difficult to find a library with the same performance than ta-lib - simply because everything is written in C, so a regular python library can only keep up if it's implemented with numpy (or similar) methods.

Therefore, you'll not find anything for candlestick-patterns i think, simply because the detection would be pretty slow in python.

**Comment by LuisAlejandro** [comment]
@xmatthias Thanks. I've decided to stick with ta-lib for a while.

**Comment by Sshimaninja** [comment]
Pretty clean Python TA library:
Title: "Technical Analysis Library in Python"
https://technical-analysis-library-in-python.readthedocs.io/en/latest/

**Comment by diatomm** [comment]
Greetings, 
I've been tinkering with this project over the weekend and I've been running into walls when it comet to the "TA issue". As I am trying to implement freqtrade on a ubuntu server via docker. This is my first experience with docker as well as freqtrade. I've tried a couple "Dockerfile.Custom" files and a few different "docker-compose.yaml" files. Now I am not really asking for a solution to the issue or an intro to docker however, I will share a couple observations I've had.

My concerns when building a custom dockerfile is that I may not be pulling all the dependencies I need vs. just pulling the image from the stable branch. i.e "requirements.txt" 

Many of the sample strats ive explored require ta and so instead of experimenting with freqtrade im spending more time debugging config files etc. 

Perhaps fleshing out a full "Quick start docker.custom file" with some tips and could be something to look at in the future. Furthermore, for new users creating more accessible strategies that are decoupled from the "TA" issue.  Lemme know what you guys think.

**Comment by xmatthias** [comment]
I'm not sure i understand your issue with the custom dockerfile.

It includes a sample library (pyti) - and it's up to you to insert the proper library you wanna use (we can't and won't provide samples for all possibilities - as there's countless ta libraries out there (most providing only few - but very specific indicators).

If the develop image is pushed or the locally available is used can be controlled by a `--pull` flag in `docker-compose build --pull` - which will always try to pull the latest upstream image before building.

**Comment by diatomm** [comment]
Yeah totally fair. I was just having issues with the yaml and docker. This is a me issue not really a ta issue. But if more sample strategies had not required to probably wouldn't have provided any feed back. Anyways thanks for the tips!

**Comment by ghost** [comment]
> Pretty clean Python TA library:
> Title: "Technical Analysis Library in Python"
> https://technical-analysis-library-in-python.readthedocs.io/en/latest/

It looks like development for this has stalled, but it looked like a good contender. Maybe worth forking or contributing to.

**Comment by xmatthias** [comment]
The best library i've (and others here) found in this regard, which is also very actively maintained is pandas-ta - which is now also included in our docker images to simplify usage.

I'll still need to include that in our documentation and maybe change some samples towards this (maybe also tests ...).

**Comment by bs32g1038** [comment]
how's the progress?

**Comment by xmatthias** [comment]
this is unlikely to be actually executed / being acted on.
While it lists some good alternatives, none has the feature set and performance of ta-lib.

as such - there's no real progress - and the most likely course of action will be that this issue will be closed at some point.

**Comment by yttasdfghjk** [comment]
In many backtests talib has not added any value as it seems to define candles differently than I would expect it to, i.e. false positives. Is removing it from core freqtrade an option still or are we gonna keep it?

**Comment by xmatthias** [comment]
unlikely still. 
I'm mainly keeping this open as a reminder that installation of it is not perfect yet. This is likely going to change as ta-lib is looking to build wheels for it's library - which (if done correctly) will remove the need for additional installation steps.

Even if it's installed though - nothing forces a strategy to use it.

**Comment by NtsakoB** [comment]
Hi everyone. How can I install the freqtrade packages one-by-one to exclude Ta-lib?

Here's my new approach:

Replace talib.abstract indicators with pandas_ta.

Use finta as a secondary fallback.

I've also added this script:
 #!/bin/bash
# Patch freqtrade to ignore TA-Lib imports
sed -i 's/import talib/# import talib/' /root/.local/lib/python3.11/site-packages/freqtrade/strategy/indicators.py

# Remove TA-Lib from freqtrade's dependencies
sed -i '/TA-Lib/d' /root/.local/lib/python3.11/site-packages/freqtrade-2023.4.0.dist-info/METADATA

This aims to comment out the import talib line & also ensure that TA-Lib is removed from freqtrade's dependencies, but it still won't stop giving me an error:

ERROR: failed to solve: process "/bin/sh -c pip install --no-cache-dir \"pip<25\" \"setuptools<70\" wheel &&     pip install --no-cache-dir --no-binary ta-lib -r requirements.txt" did not complete successfully: exit code: 1

Requirements.txt looks like this:
freqtrade==2023.4.0
pandas_ta==0.3.14b0
finta==1.3
talipp==2.5.0
numpy==1.26.0
pandas==2.0.3
python-dotenv==1.0.0
telethon==1.34.0
plotly==5.18.0
kaleido==0.2.1


Include talipp for latency-sensitive signals like pump detection and whale tracking.

All help is appreciated.

**Comment by NtsakoB** [comment]
I've also modified the Dockerfile to ensure proper installation of dependencies and avoid warnings about running pip as the root user.

**Comment by stash86** [comment]
In case you haven't notice, pandas-ta is going behind paywall now. No more free usage of newer version. So no, we aren't gonna replace ta-lib with pandas-ta nor we will provide ways to do so

**Comment by twopirllc** [comment]
Hello!

Not to advocate for the inclusion of Pandas TA into the future of **freqtrade**. But I did want to  clarify that [Pandas TA](https://www.pandas-ta.dev/) is **not** behind a paywall. In fact the latest version, _0.4.25b_, can be downloaded [here](https://www.pandas-ta.dev/getting-started/installation/#official-source).

To help cover the development and maintenance of Pandas TA, it will be transitioning to an _Insiders_ program similar to [mkdocs-material](https://squidfunk.github.io/mkdocs-material/insiders/) for access to the repo, prereleases, etc.

Keep up the good work!
KJ

**Comment by xmatthias** [comment]
@NtsakoB we're not supporting such oddly patched versions, hence i've hidden your comment as outdated to not confuse users with broken, non-supported scripts.

---

@twopirllc the wording on the (now deleted) Repository issue anouncing this was different - and in my honest understanding, newer versions of pandas-ta are no longer open source (at least i struggle to find the current source-code since the original repo was deleted), while the license seems to remain MIT. That's perfectly fine from a license perspective (in my understanding of MIT) - but that doesn't mean that we have to agree with this.

I'm not saying i suspect anything similar - but please understand that the safe way for us as a project (to protect our users from malicious future updates) is to continue going with an open source project (where the source is available) - which pandas-ta is currently not.


> In fact the latest version, 0.4.25b, can be downloaded [here](https://www.pandas-ta.dev/getting-started/installation/#official-source).

The instruction lists `pip install pandas-ta` as installation method - which installs 0.3.14b (the latest version available from pip).
This version is no longer compatible with numpy2 - so your instruction is wrong / outdated, and the pypi version is not one we'll be able to use going forward. 
Installing a "binary uploaded file from a random website" is not feasible for a project like freqtrade, with a non-neglectible userbase, where the installation happens through a `requirements.txt` file (which essentially requires dependencies to be available through pypi, the official python package index).
As the pypi version is outdated and broken with numpy2 - this makes it an effective paywall (painwall?) - which is worse than paying, as it makes installation a pain.

For Freqtrade, I've already created a patched fork release on pypi which updates numpy compatibiliy on the 0.3.x branch - which will remain open source, and to which freqtrade will update to for a transition period (removing pandas-ta immediately would break tons of user's strategies).
The main reason for this is that your issue made it clear - further issues will be closed and the github repo is no longer maintained
As I have no desire to send an email to get support from a supposed open source project - i prefer doing the 5 line patches myself.

We're however no longer linking to pandas-ta as "potential indicator library" - and have removed the import from the provided strategy templates to not encourage further development on top of pandas-ta.


---

As an aside, the whole page looks odd now - you have an empty graph in "library growth" (repo is gone - there are no stars)

![Image](https://github.com/user-attachments/assets/f553efa0-26ec-4763-8447-b01fe678570e)

This is however not the place to discuss your library's future or decisions - pandas-ta's library's issues would've been the place for this, not an issue at a random other repository that happens to use pandas-ta. 
I'm happy to discuss this further on the pandas-ta repo (or feel free to join our discord to discuss this in private) - but definitely not in this issue, so further comments talking about "future of pandas-ta" will be hidden as off-topic here.

**Comment by NtsakoB** [comment]
@xmatthias noted & all concerns considered. I found myself at a crossroads because of the the installation issue I've been experiencing. I'd much rather use freqtrade but the issue persists: note: This error originates from a subprocess, and is likely not a problem with pip.
30.47   ERROR: Failed building wheel for TA-Lib
30.47 Successfully built pandas_ta telethon py-find-1st pyaes sdnotify
30.47 Failed to build TA-Lib
30.47 ERROR: Could not build wheels for TA-Lib, which is required to install pyproject.toml-based projects
30.65 
30.65 [notice] A new release of pip is available: 24.0 -> 25.1.1
30.65 [notice] To update, run: pip install --upgrade pip

The output may be slightly different because of my recent approach but ultimately the issue is the same as with everyone.


I'v been stuck here for a month now.

**Comment by froggleston** [comment]
@NtsakoB Please open a new issue instead of necro-ing a largely unrelated issue.

**Comment by xmatthias** [comment]
@NtsakoB it's a question of "not following the docs. Freqtrade installation documents provide information on the (additional) steps necessary to successfully install ta-lib (for all 3 system variants - both in manual and automated fashion). 
Removing ta-lib won't help you any bit - as it's a dependency for most other technical analysis libraries anyway (including pandas_ta,technical, ...). So while this might make the first step seem easier - it won't be a good environment going forward.

As froggleston said, happy to assist you in your own issue - removing ta-lib from the installation will only fix installation - but you won't have a strategy you can run after that.


This is also an issue that's still here for reference (at least since https://github.com/freqtrade/freqtrade/issues/2509#issuecomment-2712694547) - to remind me to ping ta-lib-python folks from time to time to provide proper wheels (it's funny as building wheels for ta-lib ain't that hard - and we do this already for windows - but the same approach would work for other platforms, too).

I'll take this opportunity to close it and mark this task on my own todo-list - as it's pretty clear that this is confusing people more than anything.
